backend/module/audit/audit.go

```diff
@@ -52,35 +51,71 @@ func (m *mod) Register(r module.Registrar) error {
 }
 
 func (m *mod) GetEvents(ctx context.Context, req *auditv1.GetEventsRequest) (*auditv1.GetEventsResponse, error) {
+	resp := &auditv1.GetEventsResponse{}
+
+	var start time.Time
+	var end *time.Time
+
 	switch req.GetWindow().(type) {
 	case *auditv1.GetEventsRequest_Range:
 		timerange := req.GetRange()
-		start := timerange.StartTime.AsTime()
-		end := timerange.EndTime.AsTime()
-
-		events, err := m.client.ReadEvents(ctx, start, &end)
-		if err != nil {
-			return nil, err
-		}
-		return &auditv1.GetEventsResponse{Events: events}, nil
+		start = timerange.StartTime.AsTime()
+		endTime := timerange.EndTime.AsTime()
+		end = &endTime
 	case *auditv1.GetEventsRequest_Since:
 		if err := req.GetSince().CheckValid(); err != nil {
 			return nil, fmt.Errorf("problem parsing duration: %w", err)
 		}
 
 		window := req.GetSince().AsDuration()
-		start := time.Now().Add(-window)
-
-		events, err := m.client.ReadEvents(ctx, start, nil)
-		if err != nil {
-			return nil, err
-		}
-		return &auditv1.GetEventsResponse{Events: events}, nil
+		start = time.Now().Add(-window)
 	default:
 		return nil, errors.New("no time window requested")
 	}
+
+	var page int
+	if req.PageToken != "" {
+		p, err := strconv.Atoi(req.PageToken)
+		if err != nil || p < 0 {
+			return nil, fmt.Errorf("invalid page token: %s", req.PageToken)
+		}
+		page = p
+	}
+
+	limit := int(req.Limit)
+	if req.Limit == 0 {
+		limit = 1
+	}
+
+	startIdx := page * limit
+	additionalEventsNumber := int64(limit + 1)
+	options := &audit.ReadOptions{Offset: int64(startIdx)}
+	if req.Limit != 0 {
```

Comment:
is this check necessary? don't we have a default limit?

Reply:
at the moment we don't have a default. you get all the events back. In a prior commit I was defaulting to 10 but that would modify the existing behavior